src/Npgsql/NpgsqlDefaultDataReader.cs

```diff
@@ -177,11 +177,19 @@ public override T GetFieldValue<T>(int column)
 
             SeekToColumn(column);
             if (ColumnLen == -1)
-                throw new InvalidCastException("Column is null");
+            {
+                if (NullableHandler<T>.Exists)
+                    return default;
+                else
+                    throw new InvalidCastException("Column is null");
+            }
 
             var fieldDescription = RowDescription[column];
             try
             {
+                if (NullableHandler<T>.Exists)
```

Comment:
I'd still like to see this check happen only after a failure to read the value normally. Why not move this to https://github.com/npgsql/npgsql/blob/dev/src/Npgsql/TypeHandling/NpgsqlSimpleTypeHandler.cs#L122 (for simple) and to https://github.com/npgsql/npgsql/blob/dev/src/Npgsql/TypeHandling/NpgsqlTypeHandler%60.cs#L98 (for non-simple)? This would completely remove the nullable handling from the regular handling flow.

Reply:
Do you want to handle nulls in the catch block?

Reply:
No; definitely not (we shouldn't do normal flow via exception handling). But the `ReadAny<TAny>` methods on Npgsql{Simple,}TypeHandler are the place where we try to cast the handler to the interface which handles the user-requested type. If that cast fails, that's a good place for falling back and checking whether maybe the type is nullable etc. I'm still not sure if it's going to work well, but let's give it a shot.

Reply:
Does it make sense? If `len` equals to `-1` then there is no reason to use a type handler. There are no data in the buffer to read.

Reply:
Let me try to explain: I only meant that these two lines of nullable-handling might be better in `Read<Any>`, the case where `ColumnLen == -1` should probably stay exactly where it is... If it's not clear I can write some code.

Reply:
Understand you. You have said that plan to merge `NpgsqlDefaultDataReader` and `NpgsqlSequentialDataReader`. Therefore, it would be better to keep the code in the current places because after the merge the only one file will exist where `NullableHandler<T>` will be used. I want to say that it would be better to keep the class in a single file and use it only in it.

Reply:
OK, let's keep it as it is for now - we can always refactor it later.